### DISCUSSION TITLE
Is there a way to run a scenario multiple times based on the data sets specified in a yaml/json file ?

### DISCUSSION THREAD

**Discussion by ayankumaar** [getgauge/gauge#1876] [General]
I want to specify data sets in a yaml file and along with that I would like to specify the specs and scenarios within it to run. The actual data sets will be specified in a csv file. The yaml file will just be used to specify which data set to be picked. For each data set specified I would like to run each scenario once. Is there any way in gauge to achieve this ?

Here is a sample yaml file that I have created for this purpose.

TestInfo.yaml
-------------------------------------------------------------------------------------------------
command: 
  Creating Sample Spec for Automation Demo(spec): Register a user Automation Demo(scenario)
  Search for a string(spec): Search for a string in google(scenario)
dataset: 
  DS1, DS2, DS3, DS4
---------------------------------------- End of Yaml File---------------------------------------
  
My requirement is to run scenario 1- **Register a user Automation Demo**  four times (one time each with DS1,DS2,DS3,DS4)
and then run scenario 2- **Search for a string in google** four times (one time each with DS1,DS2,DS3,DS4)

My configuration is given below:
Gauge version: 1.1.4

Plugins
-------
dotnet (0.1.9)
html-report (4.0.8)
screenshot (0.0.1)

I am using gauge with C#.

I have already looked into the following issues:
#1240 and #1242. From these two I came across the concept of config.yaml file. But, didn't find any documentation or example on that. If there is a way to achieve my requirement using that then I would like to see an example on that.

**Comment by zabil** [comment]
There is no direct support for this. But you could write a script to parse the YAML to generate csv's and a command or script to run scenarios by name using `--scenario` [flag](https://manpage.gauge.org/gauge_run.html). I am afraid there's no example at the moment.

**Reply by ayankumaar** [reply]
@zabil Thank you very much for the information. Is there a way to trigger the execution of a scenario multiple times ?

**Reply by zabil** [reply]
There's no direct support for running a scenario multiple times. You can use the `--repeat` [flag](https://manpage.gauge.org/gauge_run.html) to repeat the last run.

**Reply by ayankumaar** [reply]
@zabil I found that there is a way to import a csv file in a specification like this **table: /system/users.csv**. But, is there any way to pass the csv file at run time from the gauge CLI ?

**Reply by zabil** [reply]
The csv files cannot be passed at runtime. The fields in the csv file are required for parsing Gauge specifications. Instead, as mentioned earlier, you could generate the csv data into the path specified in the specification before running gauge tests.

**Reply by ayankumaar** [reply]
@zabil I have kept a dummy csv file at a specified path. The path is specified as **table: /system/users.csv** in the specification. I am trying to fill up data into it at run time. I am using the **"BeforeSuite"** hook to fill up data into it. But, the data is not getting picked by gauge. I am noticing if the data is already present in the file before the execution starts. Only then it is getting picked. **But, any data updated at runtime from the "BeforeSuite" hook is not getting picked by gauge.**

So, what is the way to keep a dummy file and update data into it at runtime, so that the updated data is picked by gauge ?

**Reply by zabil** [reply]
This is not possible at the moment. You need to populate the csv before running the `gauge` command. You could run a script to do this and then launch `gauge` to run all the specifications.

**Reply by ayankumaar** [reply]
@zabil Can I raise a feature request for this ?

Also, I read at this link, https://github.com/getgauge/gauge/issues/586 that it is possible to set a variable in the default.properties file under the env folder. There I can mention "csv_path" and refer that path in my scenario/spec. Will that help in my case in any way ? I mean I am trying to look for a work around solution for time being. Any hints on that will be highly appreciated.

**Reply by zabil** [reply]
> Can I raise a feature request for this ?

This was requested earlier but was closed after consideration as Gauge's spec validation logic requires the csv's to be loaded before running them. In short, Gauge won't be implementing this feature because of technical difficulties.

I am not sure there's a workaround. You could read a csv file and access it's data from a step instead and add it to the [datastore](https://docs.gauge.org/writing-specifications.html?os=macos&language=javascript&ide=vscode#data-store) for other steps to work with.

**Reply by ayankumaar** [reply]
@zabil Thank you very much for all your responses. If I read a csv file and access it's data from a step and add it to the datastore for other steps to work with. Then even in that case the data needs to be filled in the csv before starting the execution ?

**Reply by zabil** [reply]
> Then even in that case the data needs to be filled in the csv before

No, because you are using custom reader to parse and read the csv from the step implementation.

**Reply by ayankumaar** [reply]
@zabil Is there an example for this ? That will help me a lot.

**Reply by zabil** [reply]
There's no example for this at the moment. A quick google points to a csv reader like https://joshclose.github.io/CsvHelper/ for c# which you could use for reading csv files.

**Reply by ayankumaar** [reply]
@zabil I am yet to find a solution for my requirement. That is why I am reaching out to you again. 

I have written few hooks to trigger execution of my gauge test cases. The hooks get executed in a particular order. For ex: BeforeSuite ->BeforeSpec -> BeforeScenario -> BeforeStep -> AfterStep -> AfterScenario -> AfterSpec -> AfterSuite. 
Now, by default these hooks get executed once per execution. 

Is there any way to execute these hooks multiple times in the same order ? If there is a way to do this then that will solve my problem.

**Reply by zabil** [reply]
> Is there any way to execute these hooks multiple times in the same order ? If there is a way to do this then that will solve my problem.

I don't think this is possible

**Reply by sriv** [reply]
Hooks get executed only once per lifecycle, (i.e BeforeSuite/AfterSuite - once per run, BeforeSpec/AfterSpec - once per spec etc). However, you can have multiple hooks, ex: you can have two `BeforeSpec` hooks.

There is no guarantee on the order of hooks, but you can use tags to filter them out for certain cases. 

In your situation, a hack would be to extract hook logic to a common method, and call that in multiple marker methods with the right annotation. ex:

```
private void commonHookStuff() {
}

@BeforeSpec
public void FirstHook() {
    commonHookStuff();
}

@BeforeSpec
public void SecondHook() {
    commonHookStuff();
}
```

But do note that in your situation, the csv files need to be present before gauge attempts it's execution lifecycle. Hence you cannot use an execution hook to populate data that would be used by gauge to parse the spec. It is important to note that Gauge's lifecycle looks roughly like this: `parse` -> `validate` -> `execute`. The hooks sit in `execute` phase, whereas the CSV files are read in the `parse` phase.
